### PR TITLE
Replace dox feature with docsrs attribute

### LINKED ISSUE #1454: Suggestion: Remove dox feature in favor of rustc args
Cargo features are supposed to be additive, but the "dox" feature isn't. Compiling with "dox" skips compile time checks for the underlying C libraries and therefore should not be enabled during normal development.

A better solution would be to use a "docsrs" or "dox" config argument for rustc directly. This seems to be what most crates do and it's also what [docs.rs recommends](https://docs.rs/about/builds#cross-compiling). This would make possible to use `--all-features` again and removes the need to pass the "dox" feature to the dependencies.

The code for conditional compilation would only change from `#[cfg(feature = "dox")]` to `#[cfg(dox)]`.

This doesn't only affect the gtk4 crate, so feel free to move the issue if there's a better repo for this topic.

### PR DESCRIPTION
**Pull request by AaronErhardt**
Fixes #1454 

I think this should be all from the gir side.
I've tested this branch on gtk4-rs and everything looked and worked great so far.

Let me know if I should submit PRs for various gtk-rs crates once this is merged.

### CODE REVIEW

File: book/src/tutorial/high_level_rust_api.md [hunk 1]
```diff
@@ -115,7 +113,7 @@ We need to include all `auto` files in our library.
 To do so, let's update the `src/lib.rs` file as follows:
 
 ```rust
-#![cfg_attr(feature = "dox", feature(doc_cfg))]
+#![cfg_attr(docsrs, feature(doc_cfg))]
```
Comment (bilelmoussaoui) [inline book/src/tutorial/high_level_rust_api.md:116]: Can't this be named dox?

**Comment by bilelmoussaoui** [comment]
I would also suggest documenting how cargo doc should be ran in the gir book if possible

File: book/src/tutorial/high_level_rust_api.md (hunk 1 above)
Comment (AaronErhardt) [reply book/src/tutorial/high_level_rust_api.md:116]: Sure, we can change this. "docsrs" is just what seems to be used in other crates, I haven't seen "dox" in the wild yet. Just let me know if you prefer "dox".

Comment (sdroege) [reply book/src/tutorial/high_level_rust_api.md:116]: Which other crates are that?

Comment (sdroege) [reply book/src/tutorial/high_level_rust_api.md:116]: I guess that decides it then. Let's use the same

File: book/src/tutorial/generate_docs.md [hunk 2]
```diff
@@ -47,6 +47,6 @@ Gir will not override them when you re-generate bindings.
 Easy, right.
 If this is your first time publishing a crate, you can find a detailed guide [here](https://doc.rust-lang.org/cargo/reference/publishing.html).
 
-Before you publish the crate, please ensure docs.rs will activate the dox feature and the dox feature of the safe wrapper crate also activates the feature of its dependencies and the unsafe FFI bindings you created.
+Before you publish the crate, please ensure docs.rs will activate the `docsrs` attribute.
```
Comment (bilelmoussaoui) [inline book/src/tutorial/generate_docs.md:50]: What about when you run cargo doc like we do on the CI side of things?

Comment (AaronErhardt) [reply book/src/tutorial/generate_docs.md:50]: If you have the gtk4-devel package installed, you should be fine without any extra effort.

For annotation of required features with `doc_attr` (e. g. for CI) you need to use nightly as usual and set one env variable for rustdoc, for example: `RUSTDOCFLAGS='--cfg docsrs' cargo +nightly doc`.

Should I mention that in the docs as well?

Comment (bilelmoussaoui) [reply book/src/tutorial/generate_docs.md:50]: Yes please

**Comment by GuillaumeGomez** [comment]
So sadly it will require to add a `--feature` argument when running `cargo doc`. Not much we can do about it... Really annoyed that `cargo` still doesn't allow to have `doc` profile for that.

**Comment by bilelmoussaoui** [comment]
> So sadly it will require to add a `--feature` argument when running `cargo doc`. Not much we can do about it... Really annoyed that `cargo` still doesn't allow to have `doc` profile for that.

why a --feature is needed? i thought everything would be handled by the RUSTFLAGS?

**Comment by GuillaumeGomez** [comment]
Sorry, you're right. Wrote too quickly. `RUSTFLAGS` is what I meant.

**Comment by sdroege** [comment]
And `generator.py` would set that accordingly before calling `cargo doc`.

**Comment by AaronErhardt** [comment]
Speaking of `generator.py`, I haven't changed anything on this file so far. As far as I understand it doesn't run `cargo doc` but expects the user to do so after following all required steps. Should I change something about that?

Everything else should be in place by now, I've pushed an update for the book earlier today.

**Comment by sdroege** [comment]
Oh right. But we set `RUSTFLAGS` for that already anyway so adding just one more thing is not a problem. Currently we have to set `RUSTFLAGS=-Z unstable-options` at least.

**Comment by AaronErhardt** [comment]
Is there still anything missing in this PR?

**Review by bilelmoussaoui — APPROVED** [review]

**Comment by sdroege** [comment]
@AaronErhardt Once merged, will you create PRs for gtk-rs-core and gtk4-rs at least to update accordingly?

**Comment by AaronErhardt** [comment]
Sure

**Comment by sdroege** [comment]
Great, then let's get this in :) I can take care of gstreamer-rs once the others are in.

**Pull request merged**